pkg/ingester/circuitbreaker.go

```diff
@@ -0,0 +1,233 @@
+		}).
+		OnOpen(func(event circuitbreaker.StateChangedEvent) {
+			circuitBreakerTransitionsCounterFn(cb.metrics, circuitbreaker.OpenState).Inc()
+			level.Warn(logger).Log("msg", "circuit breaker is open", "previous", event.OldState, "current", event.NewState)
+		}).
+		OnHalfOpen(func(event circuitbreaker.StateChangedEvent) {
+			circuitBreakerTransitionsCounterFn(cb.metrics, circuitbreaker.HalfOpenState).Inc()
+			level.Info(logger).Log("msg", "circuit breaker is half-open", "previous", event.OldState, "current", event.NewState)
+		})
+
+	if cfg.testModeEnabled {
+		// In case of testing purposes, we initialize the circuit breaker with count based failure thresholding,
+		// since it is more deterministic, and therefore it is easier to predict the outcome.
+		cbBuilder = cbBuilder.WithFailureThreshold(cfg.FailureThresholdPercentage)
+	} else {
+		// In case of production code, we prefer time based failure thresholding.
+		cbBuilder = cbBuilder.WithFailureRateThreshold(cfg.FailureThresholdPercentage, cfg.FailureExecutionThreshold, cfg.ThresholdingPeriod)
+	}
+
+	cb.cb = cbBuilder.Build()
+	cb.metrics = newCircuitBreakerMetrics(registerer, cb.cb.State)
+	return &cb
+}
+
+func isFailure(err error) bool {
+	if err == nil {
+		return false
+	}
+
+	// We only consider timeouts or ingester hitting a per-instance limit
+	// to be errors worthy of tripping the circuit breaker since these
+	// are specific to a particular ingester, not a user or request.
+
+	if errors.Is(err, context.DeadlineExceeded) {
+		return true
+	}
+
+	statusCode := grpcutil.ErrorToStatusCode(err)
+	if statusCode == codes.DeadlineExceeded {
+		return true
+	}
+
+	var ingesterErr ingesterError
+	if errors.As(err, &ingesterErr) {
+		return ingesterErr.errorCause() == mimirpb.INSTANCE_LIMIT
+	}
+
+	return false
+}
+
+func (cb *circuitBreaker) isActive() bool {
+	if cb == nil {
+		return false
+	}
+	return cb.active.Load()
+}
+
+func (cb *circuitBreaker) setActive() {
+	if cb == nil {
+		return
+	}
+	cb.active.Store(true)
+}
+
+// tryAcquirePermit tries to acquire a permit to use the circuit breaker and returns whether a permit was acquired.
+// If it was possible to acquire a permit, a status true and no error are returned. The acquired permit must be
+// returned by a call to finishPushRequest.
+// If it was not possible to acquire a permit, a status false is returned. In this case no call to finishPushRequest
+// is needed. If the permit was not acquired because of an error, that causing error is returned as well.
+func (cb *circuitBreaker) tryAcquirePermit() (bool, error) {
+	if !cb.isActive() {
+		return false, nil
+	}
+	if !cb.cb.TryAcquirePermit() {
+		cb.metrics.circuitBreakerResults.WithLabelValues(resultOpen).Inc()
+		return false, newCircuitBreakerOpenError(cb.cb.RemainingDelay())
+	}
+	return true, nil
+}
+
+func (cb *circuitBreaker) recordResult(err error) {
+	if !cb.isActive() {
+		return
+	}
+	if err != nil && isFailure(err) {
```

Comment:
Would it make sense to distinguish between "real" errors and too-slow requests in metrics?

I'm afraid it can be quite confusing to see metrics saying that there were 1000 errors, when in fact there were none, but requests were slow. If we recorded slow requests with different label, say `too_slow`, this would be clear. WDYT?

Reply:
@pstibrany I think that it is good as it is right now. In this function we define what is an error for a circuit breaker, and `deadline exceeded` errors are actually a valid errors for circuit breakers. Having an additional case would additionally complicate an already complicated implementation.
On the other hand, slow ingesters actually cause errors: the distributors calling them will report a `deadline exceeded` error, although this error might never happen on the ingester side. If we report them in cb-related metrics and record them with cbs, we would actually be more consistent with the rest of the system than now.